lib/goo/sparql/solutions_mapper.rb

```diff
@@ -157,35 +164,56 @@ def get_value_object(id, objects_new, object, list_attributes, predicate)
-                  @models_by_id[id][predicate]
-                else
-                  @models_by_id[id].instance_variable_get("@#{predicate}")
-                end
-          if object.nil? && pre.nil?
-            object = []
-          elsif object.nil? && !pre.nil?
-            object = pre
-          elsif object
-            object = !pre ? [object] : (pre.dup << object)
-            object.uniq!
-          end
+          pre = @klass_struct ? @models_by_id[id][predicate] : @models_by_id[id].instance_variable_get("@#{predicate}")
+
+          object = []  if object.nil? && pre.nil?
+
+          object = pre if object.nil? && !pre.nil?
+
+          object = pre.nil? ? [object] : (pre.dup << object)
+          object.uniq
+
         end
         [object, objects_new]
       end
 
-      def add_object_to_model(id, object, predicate, lang)
+      def object_language(new_value)
+        new_value.language || :no_lang if new_value.is_a?(RDF::Literal)
+      end
+
+      def language_match?(language, requested_lang = nil)
+        !language.nil? && (language.eql?(requested_lang) || language.eql?(:no_lang) || requested_lang.nil?)
+      end
+
+      def add_object_to_model(id, objects, current_obj, predicate, language, requested_lang = nil)
         if @models_by_id[id].respond_to?(:klass)
-          @models_by_id[id][predicate] = object unless object.nil? && !@models_by_id[id][predicate].nil?
+          @models_by_id[id][predicate] = objects unless objects.nil? && !@models_by_id[id][predicate].nil?
         elsif !@models_by_id[id].class.handler?(predicate) &&
-              !(object.nil? && !@models_by_id[id].instance_variable_get("@#{predicate}").nil?) &&
+              !(objects.nil? && !@models_by_id[id].instance_variable_get("@#{predicate}").nil?) &&
               predicate != :id
 
-          if (lang&.eql?(:no_lang)) || !lang
-            @models_by_id[id].send("#{predicate}=", object, on_load: true)
+          if language.nil? # the object is a non-literal
+            return @models_by_id[id].send("#{predicate}=", objects, on_load: true)
+          end
+
+          if language_match?(language, requested_lang) # the object is a literal and the language matches
+            return @models_by_id[id].send("#{predicate}=", objects, on_load: true)
           end
 
+
+          # the object is a literal and the language does not match , so we store it in a hash
+          @objects_by_lang[language] ||= []
+          item = @objects_by_lang[language].find { |obj| obj[:id] == id && obj[:predicate] == predicate }
+         
+          if item
+            # If an item with the matching id exists, update its attributes
+            item[:objects] << current_obj.object
+            item[:predicate] = predicate
+          else
+            # If an item with the matching id does not exist, add the new item to the array
+            @objects_by_lang[language] << { id: id, objects: [current_obj.object], predicate: predicate }
```

Comment:
is not better to have directly a hash  with this structure `hash[model_id][predicate][language] = values_array`